### ISSUE TITLE
Sorting Algorithm: Potential for failure to complete/Slow down

### ISSUE THREAD

**Issue by Taustin442** [Taustin442/polygons#22]
Out of the three times I've tested the new algorithm ( #21  ), all three tests have reached a point where one shape rapidly cycles between 2-3 points that do not complete the sort, and this seems to last for an indefined length of time. I'll do more testing one this, but the algorithm might need to be tweaked for when it is close to completetion. It also slows down when it gets close to completetion and leaves alot of empty space in the middle, but these issues aren't as important

below is an example. I'll try to narrow down what causes it later
![image](https://cloud.githubusercontent.com/assets/910711/10888952/02808cda-815d-11e5-81e9-eef0d03e32c4.png)

![image](https://cloud.githubusercontent.com/assets/910711/10888956/0866abf2-815d-11e5-981b-86cae55077c4.png)

**Comment by beechovsky** [comment]
Yeah, I noticed that. It happens occasionally with the default random
movement as well. I believe it depends on the ratios and other factors,
specifically the distribution of empty spots and the distribution of
polygons. I initially had the polygons move to the nearest empty space, and
many of them ended up trapped and therefore constantly moving. The current
algo has them move to the empty spot farthest away, and it is capable of
completing. You can toggle movement styles while the sim is running, and in
my experience last night that helped to allay the issue.

On Mon, Nov 2, 2015 at 12:27 PM, Taustin442 notifications@github.com
wrote:

> Out of the three times I've tested the new algorithm ( #21
> https://github.com/Taustin442/polygons/pull/21 ), all three tests have
> reached a point where one shape rapidly cycles between 2-3 points that do
> not complete the sort, and this seems to last for an indefined length of
> time. I'll do more testing one this, but the algorithm might need to be
> tweaked for when it is close to completetion. It also slows down when it
> gets close to completetion and leaves alot of empty space in the middle,
> but these issues aren't as important
> 
> below is an example. I'll try to narrow down what causes it later
> [image: image]
> https://cloud.githubusercontent.com/assets/910711/10888952/02808cda-815d-11e5-81e9-eef0d03e32c4.png
> 
> [image: image]
> https://cloud.githubusercontent.com/assets/910711/10888956/0866abf2-815d-11e5-981b-86cae55077c4.png
> 
> —
> Reply to this email directly or view it on GitHub
> https://github.com/Taustin442/polygons/issues/22.

## 

Jeff Bucklew

443-955-8003

bucklew.jeff@gmail.com

**Comment by Taustin442** [comment]
We may need to implement an addition to the algorithm to save itself if it fails some number of times to fit a single polygon, as in all 15 tests i've done since #21 not a single one has been able to finish a sort on its own

**Comment by beechovsky** [comment]
Try changing the movement type back to random. It may not be a problem,
since we can see the data it produces, and it is really just a type of
emergent behavior.

On Mon, Nov 2, 2015 at 12:43 PM, Taustin442 notifications@github.com
wrote:

> We may need to implement an addition to the algorithm to save itself if it
> fails some number of times to fit a single polygon, as in all 15 tests i've
> done since #21 https://github.com/Taustin442/polygons/pull/21 not a
> single one has been able to finish a sort on its own
> 
> —
> Reply to this email directly or view it on GitHub
> https://github.com/Taustin442/polygons/issues/22#issuecomment-153096506.

## 

Jeff Bucklew

443-955-8003

bucklew.jeff@gmail.com

**Comment by Taustin442** [comment]
Also after we fix this I'll test timing of it and see if and how often it is faster than the origional

**Comment by Taustin442** [comment]
I believe this qualifies as a bug, as it is obviously not intended behavior for the algorithm to get stuck in a cycle. The whole point of the "start movin" button is to sort the polygons automatically.

**Comment by beechovsky** [comment]
True enough. Let's decide to keep this here or the code thread at umbc
email. I prefer here.

I can tweak the algo, or force a change back to random at some point.

On Mon, Nov 2, 2015 at 12:45 PM, Taustin442 notifications@github.com
wrote:

> I believe this qualifies as a bug, as it is obviously not intended
> behavior for the algorithm to get stuck in a cycle. The whole point of the
> "start movin" button is to sort the polygons automatically.
> 
> —
> Reply to this email directly or view it on GitHub
> https://github.com/Taustin442/polygons/issues/22#issuecomment-153097235.

## 

Jeff Bucklew

443-955-8003

bucklew.jeff@gmail.com

**Comment by Taustin442** [comment]
My bad, for some reason I keep seeing the b in your username here and my brain thinks Bryan
I think the algorithm would benefit from some sort of save when it gets stuck, or maybe a change when it knows based on segregation that it is close to completion
No need to revert, it should be fine if we can just ensure it will complete and not get stuck cycling

**Comment by beechovsky** [comment]
I think I've already figured out a solution, but I'm at work. I'll knock it
out tonight probably.

**Comment by Taustin442** [comment]
I've transitioned to class and am on my phone so I can't check it

**Comment by chandni881** [comment]
I don't see any progress on SDP. I will try to work on it after like 4pm
today.

**Comment by JinHuiXu** [comment]
I don't see SDP is on the lecture slide schedule for the project but SSR and UIDD, and what is SSR by the way?

**Comment by JinHuiXu** [comment]
I don't see SDP is on the lecture slide schedule for the project but SSR and UIDD, and what is SSR by the way?

**Comment by beechovsky** [comment]
Let's move his convo to the 'Never Ending Sharks - Documents' thread in our
school emails.
